Mathlib/Topology/MetricSpace/CoveringNumbers.lean

```diff
@@ -0,0 +1,211 @@
+Authors: Rémy Degenne, Markus Himmel, Lorenzo Luccioli, Alessio Rondelli, Etienne Marion
+-/
+module
+
+public import Mathlib.Data.ENat.Lattice
+public import Mathlib.Data.Set.Card
+public import Mathlib.Topology.EMetricSpace.Diam
+public import Mathlib.Topology.MetricSpace.MetricSeparated
+public import Mathlib.Topology.MetricSpace.Cover
+
+/-!
+# Covering numbers
+
+We define covering numbers of sets in a pseudo-metric space, which are minimal cardinalities of
+`ε`-covers of sets. We also define the packing number, which is the maximal cardinality of
+an `ε`-separated set.
+
+We prove inequalities between these covering and packing numbers.
+
+## Main definitions
+
+* `externalCoveringNumber`: the extenal covering number of a set `A` for radius `ε` is the minimal
+  cardinal of an `ε`cover.
+* `coveringNumber`: the covering number(or internal covering number) of a set `A` for radius `ε` is
+  the minimal cardinal of an `ε`cover contained in `A`.
+* `packingNumber`: the packing number of a set `A` for radius `ε` is the maximal cardinal of
+  an `ε`-separated set in `A`.
+
+## Main statements
+
+* `externalCoveringNumber_le_coveringNumber`: the external covering number is less than or equal to
+  the covering number.
+* `packingNumber_two_le_externalCoveringNumber`: the packing number with radius `2 * ε` is
+  less than or equal to the external covering number for `ε`.
+
+## References
+
+* [R. Vershynin, *High-dimensional probability*][vershynin2018high]
+
+-/
+
+@[expose] public section
+
+open EMetric Set
+open scoped ENNReal NNReal
+
+namespace Metric
+
+variable {X : Type*} [PseudoEMetricSpace X] {A B C : Set X} {ε δ : ℝ≥0} {x : X}
+
+section Definitions
+
+/-- The extenal covering number of a set `A` in `X` for radius `ε` is the minimal cardinal of
+an `ε`cover by points in `X` (not necessarily in `A`). -/
+noncomputable
+def externalCoveringNumber (ε : ℝ≥0) (A : Set X) : ℕ∞ :=
+  ⨅ (C : Set X) (_ : IsCover ε A C), C.encard
+
+/-- The covering number(or internal covering number) of a set `A` for radius `ε` is
+the minimal cardinal of an `ε`cover contained in `A`. -/
+noncomputable
+def coveringNumber (ε : ℝ≥0) (A : Set X) : ℕ∞ :=
+  ⨅ (C : Set X) (_ : C ⊆ A) (_ : IsCover ε A C), C.encard
+
+/-- The packing number of a set `A` for radius `ε` is the maximal cardinal of an `ε`-separated set
+in `A`. -/
+noncomputable
+def packingNumber (ε : ℝ≥0) (A : Set X) : ℕ∞ :=
+  ⨆ (C : Set X) (_ : C ⊆ A) (_ : IsSeparated ε C), C.encard
+
+end Definitions
+
+@[simp]
+lemma externalCoveringNumber_empty (ε : ℝ≥0) : externalCoveringNumber ε (∅ : Set X) = 0 := by
+  simp [externalCoveringNumber]
+
+@[simp]
+lemma coveringNumber_empty (ε : ℝ≥0) : coveringNumber ε (∅ : Set X) = 0 := by simp [coveringNumber]
+
+@[simp]
+lemma externalCoveringNumber_eq_zero :
+    externalCoveringNumber ε A = 0 ↔ A = ∅ := by simp [externalCoveringNumber]
+
+@[simp]
+lemma coveringNumber_eq_zero : coveringNumber ε A = 0 ↔ A = ∅ := by simp [coveringNumber]
```

Comment:
I think you should have the `Set.Nonempty` versions of these too. You could use them a bit later.